src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/AssetCacheEntry.cs

```diff
@@ -1,11 +1,81 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License. See LICENSE in the project root for license information.
 
+using System;
+using UnityEngine;
+
 namespace Microsoft.MixedReality.SpectatorView
 {
+    [Serializable]
+    internal class AssetId
+    {
+        public static AssetId Empty { get; } = new AssetId(System.Guid.Empty, -1);
+
+        /// <summary>
+        /// The Unity guid for the Asset
+        /// </summary>
+        public StringGuid Guid => guid;
+
+        [SerializeField]
+        private StringGuid guid;
+
+        /// <summary>
+        /// The Unity file identifier for the Asset
+        /// </summary>
+        public long FileIdentifier => fileIdentifier;
+
+        [SerializeField]
+        private long fileIdentifier;
+
+        public AssetId(StringGuid guid, long fileIdentifier)
+        {
+            this.guid = guid;
+            this.fileIdentifier = fileIdentifier;
+        }
+
+        public override bool Equals(object obj)
+        {
+            AssetId assetId = obj as AssetId;
+            if (assetId == null)
+            {
+                return false;
+            }
+
+            return this == assetId;
+        }
+
+        public override int GetHashCode()
+        {
+            return FileIdentifier.GetHashCode() ^ Guid.ToString().GetHashCode();
+        }
+
+        public static bool operator ==(AssetId lhs, AssetId rhs)
+        {

```

Comment:
I would simplify this to: Equals(lhs, rhs); which would propogate to your Equals override above.

I would update that method to not use this one, as you already did a null check, and "this" can't be null. So just compare Guid and FileIndentierifer.

Also, don't do Guid.ToString() for comparison, just compare Guids. Guid.Tostring() is an allocation

Reply:
Sounds good. I'll pick these changes up in stabilization/1.1.0 with the expectation to merge back into master

Reply:
hm, what aspect of Guid.ToString() is an allocation? StringGuid has a custom ToString definition that just returns the internally stored string value for the Guid. I'm not 100% if historically this type was made for serializtion, but we aren't currently using the default Guid type.

Reply:
```
        public override string ToString()
        {
            return (m_storage == null) ? System.Guid.Empty.ToString("D") : m_storage;
        }
```
Now looking at:
https://github.com/microsoft/referencesource/blob/master/mscorlib/system/guid.cs

```
            if (formatCh == 'D' || formatCh == 'd') {
                guidString = string.FastAllocateString(36);
            }
```
Looks like there is no special-casing for an empty guid, unless I missed it.